lib/lower/ASTBuilder.cpp

```diff
@@ -0,0 +1,783 @@
+    con.subs = new std::list<AST::Evaluation>();
+    pushEval(con.subs);
+    parents.emplace_back(&con);
+    return true;
+  }
+
+  void exitConstruct() {
+    popEval();
+    parents.pop_back();
+  }
+
+  // When we enter a module structure, we want to build a new module and
+  // set the builder's function cursor to point to it.
+  template <typename A>
+  bool enterModule(const A &f) {
+    auto &unit = addUnit(AST::ModuleLikeUnit{f, parents.back()});
+    funclist = &unit.funcs;
+    parents.emplace_back(&unit);
+    return true;
+  }
+
+  void exitModule() {
+    funclist = nullptr;
+    parents.pop_back();
+  }
+
+  template <typename A>
+  A &addUnit(const A &unit) {
+    pgm->getUnits().emplace_back(unit);
+    return std::get<A>(pgm->getUnits().back());
+  }
+
+  template <typename A>
+  A &addFunc(const A &func) {
+    if (funclist) {
+      funclist->emplace_back(func);
+      return funclist->back();
+    }
+    return addUnit(func);
+  }
+
+  /// move the Evaluation to the end of the current list
+  AST::Evaluation &addEval(AST::Evaluation &&eval) {
+    assert(funclist && "not in a function");
+    assert(evallist.size() > 0);
+    evallist.back()->emplace_back(std::move(eval));
+    return evallist.back()->back();
+  }
+
+  /// push a new list on the stack of Evaluation lists
+  void pushEval(std::list<AST::Evaluation> *eval) {
+    assert(funclist && "not in a function");
+    assert(eval && eval->empty() && "evaluation list isn't correct");
+    evallist.emplace_back(eval);
+  }
+
+  /// pop the current list and return to the last Evaluation list
+  void popEval() {
+    assert(funclist && "not in a function");
+    evallist.pop_back();
+  }
+
+  AST::Program *pgm;
+  std::list<AST::FunctionLikeUnit> *funclist{nullptr};
+  std::vector<std::list<AST::Evaluation> *> evallist;
+  std::vector<AST::ParentType> parents;
+};
+
+template <typename A>
+constexpr bool hasErrLabel(const A &stmt) {
+  if constexpr (std::is_same_v<A, Pa::ReadStmt> ||
+                std::is_same_v<A, Pa::WriteStmt>) {
+    for (const auto &control : stmt.controls) {
+      if (std::holds_alternative<Pa::ErrLabel>(control.u)) {
+        return true;
+      }
+    }
+  }
+  if constexpr (std::is_same_v<A, Pa::WaitStmt> ||
+                std::is_same_v<A, Pa::OpenStmt> ||
+                std::is_same_v<A, Pa::CloseStmt> ||
+                std::is_same_v<A, Pa::BackspaceStmt> ||
+                std::is_same_v<A, Pa::EndfileStmt> ||
+                std::is_same_v<A, Pa::RewindStmt> ||
+                std::is_same_v<A, Pa::FlushStmt>) {
+    for (const auto &spec : stmt.v) {
+      if (std::holds_alternative<Pa::ErrLabel>(spec.u)) {
+        return true;
+      }
+    }
+  }
+  if constexpr (std::is_same_v<A, Pa::InquireStmt>) {
+    for (const auto &spec : std::get<std::list<Pa::InquireSpec>>(stmt.u)) {
+      if (std::holds_alternative<Pa::ErrLabel>(spec.u)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+template <typename A>
```

Comment:
Could be combined with the previous and following template member function by generalizing over the types (`ErrLabel`, `EorLabel`, &c.) of the control list specifiers.

Reply:
Same answer as for other comments about combining code with templates, I agree they could, I think it also makes them easier to reason about right now.

Reply:
I disagree; a template would ease reasoning about the code by making it clear that all these cases have identical handling.

Reply:
Eventually templatized as suggested.